oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,208 @@
+
+***********************
+Translations Management
+***********************
+
+Message Ids
+===========
+
+Message ids, or message keys, uniquely identify a string to be internationalized. They should consist of one or more words to namespace the message, plus one or more words to identify the message. All words will be lowercase and separated by periods. Here are examples of message ids used in a header and in a login modal, respectively:
+
+  | ``header.search``
+  | ``login.no.such.password``
+  | ``login.privacy.policy``
+
+The part of the id used for namespacing is purely to reduce collisions and help human readability; it's not intended to be programmatically parsable.
+
+- add a stronger rec. here! translations shouldn't conflict
+
+Alternatives Considered
+-----------------------
+
+Previous i18n implementations have sometimes used the English text as the message id. While this saves developers a step in their work, it has the drawback that a change in the English copy invalidates all existing translations, even if they would not otherwise have needed to change. (For example, capitalizing a noun in English has no impact on the German translation, because nouns are capitalized in German anyway, or an English change might be correcting a typo that was correct in the translations.) Do not use English copy as the message id.
+
+XSS Issues
+==========
+
+**[TODO]**
+
+Missing Translations
+====================
+
+Missing translations can be a disorienting experience for a user. We strongly suggest a careful evaluation of translation coverage, or a clear indicator of beta status, before making a new language available to the public.
+
+****************
+Server-Side i18n
+****************
+
+**[TODO]**
+
+#################
+Technology Choice
+#################
+
+**********
+react-intl
+**********
+
+We will use `react-intl <https://github.com/yahoo/react-intl>`_ to internationalize React components, in agreement with `Paragon ADR-2 <https://github.com/edx/paragon/blob/master/docs/decisions/0002-react-i18n.rst>`_. react-intl is in line with current industry standards in i18n and satisfies our currently identified i18n needs, including multiple plurals. In order to add some browser shims, the Architecture team has published a thin wrapper around react-intl in the `frontend-i18n repo <https://github.com/edx/frontend-i18n>`_. For consistency, use this wrapper, instead of importing react-intl directly.
+
+Alternatives Considered
+=======================
+
+An alternative choice we considered was `i18next <https://react.i18next.com/>`_. It comes with plug-and-play functionality to determine a user's locale, babel scripts for various translation management schemes, etc. However, it is fairly large, and doesn't seem to offer compelling benefits that would outweigh the fact that studio-frontend and Paragon already use react-intl.
```

Comment:
Paragon doesn't have an i18n library FYI.

Reply:
D'oh!  I think that was just where the original react-intl ADR was.  Thanks!